File: common/src/main/java/com/terrasect/common/generation/Region.java
package com.terrasect.common.generation;

import com.terrasect.common.generation.definition.RegionDefinition;

import java.util.Collections;
import java.util.List;
import java.util.Set;

public record Region(
    String name,
    int areaBudget,
    RegionDefinition definition,
    Set<String> adjacentTo,
    List<Region> children
) {
    public Region {
        if (definition == null) definition = RegionDefinition.empty();
        if (adjacentTo == null) adjacentTo = Collections.emptySet();
        if (children == null) children = Collections.emptyList();
    }

    public boolean hasChildren() {
        return !children.isEmpty();
    }
}

File: common/src/main/kotlin/terrasect/definition/Region.kt
package terrasect.definition

import terrasect.sdf.Decoration

class Region(
  val name: String,
  val budget: Long,
  val children: Set<Region>,
  val strategy: Strategy? = null,
  val decorations: List<Decoration> = emptyList(),
  val originAnchor: Boolean = false,
  val climate: ClimateConstraints? = null,
  val height: HeightConstraints? = null,
  val noise: NoiseConstraints? = null,
  val biomes: SelectionConstraints? = null,
  val structures: StructureConstraints? = null,
  val mobs: SelectionConstraints? = null,
  val loot: SelectionConstraints? = null,
) {
  val bytes = name.toByteArray()
  val hasChildren = children.isNotEmpty()

  companion object {
    fun empty(name: String): Region = Region(name, 10000, emptySet())
  }
}
